### PR TITLE
Use tokio::task::spawn_blocking when generating attachment thumbnails

### PR DESCRIPTION
**Pull request by gferon**
I realized that the executor stalled when calling `send_attachment`. It turns out there's some blocking code when the `image-proc` feature is enabled with thumbnail generation and using `tokio::task::spawn_blocking` helps.

The only nice way I could find to do this is to copy the data, which is far from ideal.

### CODE REVIEW

**Comment by jplatte** [comment]
Maybe we should change the `data` parameter there frmo `&[u8]` to `Vec<u8>`? Wdyt @matrix-org/rust?

**Comment by poljar** [comment]
> Maybe we should change the `data` parameter there from `&[u8]` to `Vec<u8>`? Wdyt @matrix-org/rust?

I'm fine with changing the parameter but doesn't this break the WASM target?

**Comment by jplatte** [comment]
Does `image` work on wasm? We could still feature-gate the spawning in that case.

**Comment by poljar** [comment]
> Does `image` work on wasm? We could still feature-gate the spawning in that case.

It does, not sure how well but I've seen WASM examples using the `image` crate.

**Comment by gferon** [comment]
Yes, perfect, thanks!

**Comment by gferon** [comment]
@jplatte the last changeset should be good enough. I had to play around with returning `(data, res)` to avoid using an `Arc`.

**Comment by jplatte** [comment]
Hm, this is failing on checks that were fixed on main. I'll close and re-open but if that doesn't fix things, I think this needs a rebase.

**Review by jplatte — APPROVED** [review]

**Review by jplatte — CHANGES_REQUESTED** [review]
Wait no, this still unconditionally uses tokio. It shouldn't be used on `#[cfg(target_arch = "wasm32")]`.

**Comment by gferon** [comment]
All checks have passed on this PR, do you think it's worth merging @jplatte or should I close it?

**Comment by jplatte** [comment]
Oh sorry, this is definitely still something we want, but I need to do another review.

File: crates/matrix-sdk/src/room/joined.rs [hunk 1]
```diff
@@ -686,12 +686,32 @@ impl Joined {
...
+            let (data, thumbnail) = if config.generate_thumbnail {
+                let content_type = content_type.clone();
+
+                #[cfg(not(target_arch = "wasm32"))]
+                let (data, res) = tokio::task::spawn_blocking(move || {
+                    let res = generate_image_thumbnail(
+                        &content_type,
+                        Cursor::new(&data),
+                        config.thumbnail_size,
+                    );
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/joined.rs:698]: Instead of copy-pasting this, can you define a closure like

```rust
let make_thumbnail = move |data| { ... };

#[cfg(not(target_arch = "wasm32"))]
let (data, res) = tokio::task::spawn_blocking(make_thumbnail(data)).await.expect("Task join error");

#[cfg(target_arch = "wasm32")]
let (data, res) = make_thumbnail(data);
```

File: crates/matrix-sdk/tests/integration/room/joined.rs [hunk 2]
```diff
@@ -310,7 +310,7 @@ async fn room_attachment_send() {
         .send_attachment(
             "image",
             &mime::IMAGE_JPEG,
-            b"Hello world".to_vec(),
+            "Hello world".as_bytes().to_vec(),
```
Comment (jplatte) [inline crates/matrix-sdk/tests/integration/room/joined.rs:313]: Can you please revert the unrelated changes in this file?

File: examples/image_bot/src/main.rs [hunk 3]
```diff
@@ -16,7 +16,7 @@ async fn on_room_message(event: OriginalSyncRoomMessageEvent, room: Room, image:
 
     if text_content.body.contains("!image") {
         println!("sending image");
-        room.send_attachment("cat", &mime::IMAGE_JPEG, image, AttachmentConfig::new())
+        room.send_attachment("cat", &mime::IMAGE_JPEG, image.to_vec(), AttachmentConfig::new())
```
Comment (jplatte) [inline examples/image_bot/src/main.rs:19]: Same here.

File: crates/matrix-sdk/src/room/joined.rs [hunk 4]
```diff
@@ -686,12 +686,25 @@ impl Joined {
...
+                    (data, res)
+                };
+
+                #[cfg(not(target_arch = "wasm32"))]
+                let (data, res) = tokio::task::spawn_blocking(make_thumbnail(data))
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/joined.rs:700]: Ah sorry, has to be

```suggestion
                let (data, res) = tokio::task::spawn_blocking(move || make_thumbnail(data))
```

**Comment by jplatte** [comment]
Can you have a look at this error? I think we'll have to clone the content type, unrelated to whether there's a local closure or not (IIRC CI wasn't green, it just didn't run at all, before my latest review right?)

**Comment by gferon** [comment]
> Can you have a look at this error? I think we'll have to clone the content type, unrelated to whether there's a local closure or not (IIRC CI wasn't green, it just didn't run at all, before my latest review right?)

It's a little bit weird, since `cargo check` passes but `cargo doc` doesn't. I'll have a look.

**Comment by gferon** [comment]
@jplatte looks like the code coverage CI job is broken, I just merged the `main` branch in case it was fixed, but I believe this is good to go now!

**Comment by jplatte** [comment]
Yeah, that was an old issue that has been fixed on main. The next CI run should succeed. Thanks for pulling through with all the required changes!

**Comment by jplatte** [comment]
Unreliable codecov bot was blocking auto-merge 😒
